ldp/nn/agent/simple_local_agent.py

```diff
@@ -0,0 +1,149 @@
+from typing import cast
+
+import torch
+import torch.distributed as dist
+from aviary.core import Message, Tool, ToolRequestMessage
+from aviary_internal.nn.chunking import TensorChunker
+
+from ldp.agent import Agent, SimpleAgentState
+from ldp.graph import OpResult
+from ldp.graph.op_utils import compute_graph
+from ldp.llms import prepend_sys
+
+from ..graph.llm_call_op import LocalLLMCallOp  # noqa: TID252
+from ..handlers.transformer_handler import (  # noqa: TID252
+    ParallelModeConfig,
+    logits_to_logprobs,
+)
+from ..lm_config import LMConfig as _LMConfig  # noqa: TID252
+
+
+class LMConfig(_LMConfig):
+    """Adds some additional configuration options for running an LM in an Op."""
+
+    # async batching parameters
+    batch_size: int = 1
+    max_wait_interval: float = 0.1
+
+    # distribution
+    parallel_mode: ParallelModeConfig | None = None
+
+    # sampling parameters
+    temperature: float = 1.0
+    max_new_tokens: int = 50
+    parallel_tool_calls: bool = False
+
+
+# TODO: consider merging with SimpleAgent
+class SimpleLocalLLMAgent(Agent[SimpleAgentState]):
+    """Simple agent that can pick and invoke tools with a local language model."""
+
+    def __init__(
+        self,
+        llm_model: LMConfig,
+        sys_prompt: str | None = None,
+        hide_old_env_states: bool = False,
+    ):
+        self.llm_model = llm_model
+        self.sys_prompt = sys_prompt
+        self.hide_old_env_states = hide_old_env_states
+
+        # Initialize the local LLM operation with configuration specifics
+        self._llm_call_op = LocalLLMCallOp(
+            model_config=self.llm_model,
+            batch_size=self.llm_model.batch_size,
+            max_wait_interval=self.llm_model.max_wait_interval,
+            parallel_mode_config=self.llm_model.parallel_mode,
+        )
+
+    async def init_state(self, tools: list[Tool]) -> SimpleAgentState:
+        """Initialize the state with provided tools."""
+        return SimpleAgentState(
+            tools=tools, hide_old_env_states=self.hide_old_env_states
+        )
+
+    @compute_graph()
+    async def get_asv(
+        self, agent_state: SimpleAgentState, obs: list[Message]
+    ) -> tuple[OpResult[ToolRequestMessage], SimpleAgentState, float]:
+        """Generates an action-state-value tuple using the local LLM model."""
+        next_state = agent_state.get_next_state(obs)
+
+        # If system prompt exists, prepend it; otherwise, use the existing messages
+        messages = (
+            prepend_sys(next_state.messages, sys_content=self.sys_prompt)
+            if self.sys_prompt is not None
+            else next_state.messages
+        )
+
+        # Execute the LLM operation call
+        result = cast(
+            OpResult,
```

Comment:
```suggestion
            OpResult[Message | ToolRequestMessage],
```

This adds context that helps the below `isinstance` check make sense